src/Psl/IO/Reader.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\IO;
+
+use Psl;
+use Psl\Str;
+
+final class Reader implements ReadHandle
+{
+    private ReadHandle $handle;
+
+    private bool    $eof = false;
+    private string  $buffer = '';
+
+    public function __construct(ReadHandle $handle)
+    {
+        $this->handle = $handle;
+    }
+
+    public function getHandle(): ReadHandle
```

Comment:
Maybe better not to provide the `ReadHandle` again from the reader?
You already have the handle whilst constructing if you really want to do something with it in-between reading.

Reply:
I think it's better to keep this, `Reader` is buffered, this is useful if you want to access the un-buffered handle directly.